numpy/_core/src/umath/string_buffer.h

```diff
@@ -1428,5 +1436,73 @@ string_replace(Buffer<enc> buf1, Buffer<enc> buf2, Buffer<enc> buf3, npy_int64 c
 }
 
 
+template <ENCODING enc>
+static inline npy_intp
+string_expandtabs_length(Buffer<enc> buf, npy_int64 tabsize)
+{
+    size_t len = buf.num_codepoints();
+
+    npy_intp new_len = 0, line_pos = 0;
+
+    Buffer<enc> tmp = buf;
+    for (size_t i = 0; i < len; i++) {
+        npy_ucs4 ch = *tmp;
+        if (ch == '\t') {
+            if (tabsize > 0) {
+                npy_intp incr = tabsize - (line_pos % tabsize);
+                line_pos += incr;
+                new_len += incr;
+            }
+        }
+        else {
+            line_pos += 1;
+            size_t n_bytes = tmp.num_bytes_next_character();
+            new_len += n_bytes;
+            if (ch == '\n' || ch == '\r') {
+                line_pos = 0;
+            }
+        }
+        if (new_len > PY_SSIZE_T_MAX - 1) {
+            npy_gil_error(PyExc_OverflowError, "new string is too long");
+            return -1;
+        }
```

Comment:
@ngoldbaum Wouldn't this overflow if we have a string where there's a tab right before reaching `PY_SSIZE_T_MAX`? This might lead to an incr of more than one, which means that `new_len` will turn negative. Am I missing something?

Reply:
Thanks for pointing that out! Is checking if `new_len` is equal to `Py_SSIZE_T_MAX or negative sufficient? I just pushed a fixup.

Constructing such a string on a 64 bit system isn't feasible so it's not straightforward to test this.

Reply:
That is okay, yes. Let's merge this!